File: leetcode/872/solution.go
/*
*
https://leetcode.com/problems/leaf-similar-trees/description/?envType=daily-question&envId=2024-01-05
872. Leaf-Similar Trees
*/
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	arr1 := dfs(root1)
	arr2 := dfs(root2)
	if len(arr1) != len(arr2) {
		return false
	}
	for i, n := range arr1 {
		if n != arr2[i] {
			return false
		}
	}
	return true
}

func dfs(node *TreeNode) []int {
	if node == nil {
		return []int{}
	}
	if node.Left == nil && node.Right == nil {
		return []int{node.Val}
	}
	arr := make([]int, 0)
	if node.Left != nil {
		arr = append(arr, dfs(node.Left)...)
	}
	if node.Right != nil {
		arr = append(arr, dfs(node.Right)...)
	}
	return arr
}

File: leetcode/872/solution.py
"""
https://leetcode.com/problems/leaf-similar-trees/description/?envType=daily-question&envId=2024-01-05
872. Leaf-Similar Trees
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        return dfs(root1) == dfs(root2)


def dfs(node: Optional[TreeNode]) -> List[int]:
    if node is None:
        return []
    if node.left is None and node.right is None:
        return [node.val]
    arr = []
    if node.left:
        arr += dfs(node.left)
    if node.right:
        arr += dfs(node.right)
    return arr
